docs/proposals/env-vars.md

```diff
@@ -0,0 +1,316 @@
+	// +optional
+	EnvFrom []EnvFromSource `json:"envFrom,omitempty" protobuf:"bytes,19,rep,name=envFrom"`
+	// List of environment variables to set in the container.
+	// Cannot be updated.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Env []EnvVar `json:"env,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,7,rep,name=env"`
+```
+
+First decision:  do we need to expose both `EnvFrom` and `EnvVar` off of `Containers`, especially given that
+`EnvVar` has an optional `EnvFromSource` [ref](https://github.com/kubernetes/api/blob/v0.21.0/core/v1/types.go#L1929)
+
+```go
+// EnvVar represents an environment variable present in a Container.
+type EnvVar struct {
+	// Name of the environment variable. Must be a C_IDENTIFIER.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// Optional: no more than one of the following may be specified.
+
+	// Variable references $(VAR_NAME) are expanded
+	// using the previous defined environment variables in the container and
+	// any service environment variables. If a variable cannot be resolved,
+	// the reference in the input string will be unchanged. The $(VAR_NAME)
+	// syntax can be escaped with a double $$, ie: $$(VAR_NAME). Escaped
+	// references will never be expanded, regardless of whether the variable
+	// exists or not.
+	// Defaults to "".
+	// +optional
+	Value string `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	// Source for the environment variable's value. Cannot be used if value is not empty.
+	// +optional
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.
+
+Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
+`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  
+
+With that, a new Shipwright type would be:
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
+   // after any Parameter variable substitution.
+   // +optional
+   DefaultValue  string
+}
+```
+
+Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 
+updated with an array of `EnvironmentVariable`
+
+`BuildStrategy` would also have a `EnvironmentVariable` array named `DisallowedEnvironmentVariables` that lists `EnvVar` settings which are not allowed
+and will be filtered in the final application onto the `TaskSpec`.
```

Comment:
Without having deeply thought about this ...

* black listing might also be something on the deployment level.
* white listing might be interesting as well, for example in Paketo, environment variables are typically prefixed with `BP`

You may also add some words about what happens if a Build or BuildRun contains some environment variables that are not allowed.

Reply:
ah yes I only implied here that filtering would occur from the unallowed list

do we update status/conditions?
do we add logs / events?
do we prevent execution (seems extreme ... but maybe an opt in for the administrator)?

I'll ponder and update, thanks.  Of course chime in with any immediate opinions on how to answer those questions.

Reply:
I think the behavior should be the same as for params, there I had this exchange with Enrique: https://github.com/shipwright-io/build/pull/697/files#r604954782. So, I in general prefer that we do not just silently filter. I consider writing to the controller's container log or events silently because that's not where a build user would necessarily look at. I would prefer the validation to happen as early as possible - with webhooks, this would means to prevent the creation of the build or buildrun. With our current approach, their status would be not registered or not succeeded with reasonable reasons. Anyway, when we create the TaskRun, we always need to validate again because the build strategy might have changed after a build was created (and successfully validated by an eventually existing webhook or registered by today's approach).

Reply:
I need to re-review the merged parameters EP in details, but on the surface, I'm good with being consistent and updating this EP as needeed for consistency.  

+1 on non-silent filtering, the earlier the better, all that.

@SaschaSchwarze0 and I can dive into this more when we chat.